src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c

```diff
@@ -0,0 +1,125 @@
+// Copyright 2018 eSOL Co.,Ltd.
+// Copyright 2018 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "udp_transport_datagram_internal.h"
+#include "arpa/inet.h"
+#include "sys/socket.h"
+#include "netinet/in.h"
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+
+bool uxr_init_udp_transport_datagram(
+        uxrUDPTransportDatagram* transport)
+{
+    bool rv = false;
+
+    transport->fd = socket(AF_INET, SOCK_DGRAM, 0);
+
+    if (-1 != transport->fd)
+    {
+        /* Poll setup. */
+        #pragma GCC diagnostic push
+        #pragma GCC diagnostic ignored "-Wsign-conversion"
+        FD_ZERO(&transport->select_fd);
+        FD_SET(transport->fd, &transport->select_fd);
+        #pragma GCC diagnostic pop
+        rv = true;
+    }
+
+    return rv;
+}
+
+bool uxr_close_udp_transport_datagram(
+        uxrUDPTransportDatagram* transport)
+{
+    (void) close(transport->fd);
+
+    return true;
+}
+
+bool uxr_udp_send_datagram_to(
+        uxrUDPTransportDatagram* transport,
+        const uint8_t* buf,
+        size_t len,
+        const TransportLocator* locator)
+{
+    bool rv = true;
+
+    struct sockaddr_in remote_addr;
+    memcpy(&remote_addr.sin_addr.s_addr, &locator->_.medium_locator.address, sizeof(locator->_.medium_locator.address));
+    if (0 != remote_addr.sin_addr.s_addr)
+    {
+        remote_addr.sin_family = AF_INET;
+        remote_addr.sin_port = htons(locator->_.medium_locator.locator_port);
+
+        int32_t bytes_sent = sendto(transport->fd, (void*)buf, len, 0,
+                        (struct sockaddr*)&remote_addr, sizeof(remote_addr));
+
+        if (0 >= bytes_sent)
+        {
+            rv = false;
+        }
+    }
+
+    return rv;
+}
+
+bool uxr_udp_recv_datagram(
+        uxrUDPTransportDatagram* transport,
+        uint8_t** buf,
+        size_t* len,
+        int timeout)
+{
+    bool rv = false;
+
+    struct timeval tv;
+    tv.tv_sec = timeout / 1000;
+    tv.tv_usec = (timeout % 1000) * 1000;
+
+
+    int32_t poll_rv = select(transport->fd + 1, &transport->select_fd, NULL, NULL, &tv);
```

Comment:
Bug: The same FD_ZERO FD_SET treatment before select() should be done here. Will post a new PR soon.

Reply:
Cool, thanks. We will handle it in a new PR.

Reply:
Yes! We missed that the set object is modified.

We also use a select approach on the multi-serial agent, where it is enough to copy the set on a local variable, check here: [link](https://github.com/eProsima/Micro-XRCE-DDS-Agent/blob/develop/src/cpp/transport/serial/MultiSerialAgentLinux.cpp#L82) (`read_fds` is the equivalent to your `transport->select_fd`).

With this approach you avoid modifications on external data.

Reply:
I like it, cleaner! I will also then update the other two usages of select() like this.